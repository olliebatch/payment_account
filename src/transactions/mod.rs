mod read;
use serde::Deserialize;

pub use read::*;

#[derive(serde::Serialize, serde::Deserialize, Clone, Copy, Debug, Eq, PartialEq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum TransactionType {
    Deposit,
    Withdrawal,
    Dispute,
    Resolve,
    Chargeback,
}

#[derive(Deserialize, Debug)]
struct Transaction {
    payment_type: TransactionType,
    client: u16,
    tx: u32,
    amount: f32,
}
#[derive(Deserialize, Debug)]
pub struct TransactionDetails {
    payment_type: TransactionType,
    tx: u32,
    amount: f32,
}

impl From<Transaction> for TransactionDetails {
    fn from(transaction: Transaction) -> Self {
        TransactionDetails {
            payment_type: transaction.payment_type,
            tx: transaction.tx,
            amount: transaction.amount,
        }
    }
}
