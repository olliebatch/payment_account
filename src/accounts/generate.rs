use crate::accounts::ClientAccount;
use crate::transactions::{Transaction, TransactionType};
use std::collections::HashMap;

pub fn generate_client_accounts(transactions: Vec<Transaction>) -> HashMap<u16, ClientAccount> {
    let mut client_accounts: HashMap<u16, ClientAccount> = HashMap::new();

    for transaction in transactions {
        let rounded_transaction = transaction.round();
        let account = client_accounts
            .entry(rounded_transaction.client)
            .or_insert(ClientAccount::new(rounded_transaction.client));

        let client = rounded_transaction.client;

        if transaction.payment_type == TransactionType::Deposit {
            let updated_account = account.clone().deposit_to_account(rounded_transaction);
            client_accounts.insert(client, updated_account);
        } else if transaction.payment_type == TransactionType::Withdrawal {
            let updated_account = account.clone().withdraw_from_account(rounded_transaction);
            client_accounts.insert(client, updated_account);
        } else if transaction.payment_type == TransactionType::Dispute {
            let updated_account = account.clone().dispute_to_account(rounded_transaction);
            client_accounts.insert(client, updated_account);
        } else if transaction.payment_type == TransactionType::Resolve {
            let updated_account = account.clone().resolve_to_account(rounded_transaction);
            client_accounts.insert(client, updated_account);
        } else if transaction.payment_type == TransactionType::Chargeback {
            let updated_account = account.clone().chargeback_to_account(rounded_transaction);
            client_accounts.insert(client, updated_account);
        }
    }
    client_accounts
}

#[cfg(test)]
mod tests {

    use crate::accounts::generate_client_accounts;
    use crate::transactions::Transaction;
    use rstest::rstest;
    use rust_decimal_macros::dec;

    #[rstest()]
    fn test_basic() {
        let mut transactions = vec![];
        transactions.push(Transaction::new_deposit(1, 1, Some(dec!(1.0))));
        transactions.push(Transaction::new_deposit(2, 2, Some(dec!(2.0))));
        transactions.push(Transaction::new_deposit(1, 3, Some(dec!(2.0))));
        transactions.push(Transaction::new_withdrawal(1, 4, Some(dec!(1.5))));
        transactions.push(Transaction::new_withdrawal(2, 5, Some(dec!(3.0))));

        let result = generate_client_accounts(transactions);

        insta::assert_json_snapshot!(serde_json::json!(result));
    }

    #[rstest()]
    fn test_extra_decimal_points() {
        let mut transactions = vec![];
        transactions.push(Transaction::new_deposit(1, 1, Some(dec!(1.12345678))));
        transactions.push(Transaction::new_deposit(2, 2, Some(dec!(2.12345678))));
        transactions.push(Transaction::new_deposit(1, 3, Some(dec!(2.12345678))));
        transactions.push(Transaction::new_withdrawal(1, 4, Some(dec!(2.12345678))));
        transactions.push(Transaction::new_withdrawal(2, 5, Some(dec!(2.12345678))));

        let result = generate_client_accounts(transactions);

        insta::assert_json_snapshot!(serde_json::json!(result));
    }

    #[rstest(
    transactions,
    case::unresolved_dispute(
        vec![Transaction::new_deposit(1, 1, Some(dec!(1.0))),
        Transaction::new_deposit(2, 2, Some(dec!(2.0))),
        Transaction::new_deposit(1, 3, Some(dec!(2.0))),
        Transaction::new_withdrawal(1, 4, Some(dec!(1.5))),
        Transaction::new_withdrawal(2, 5, Some(dec!(3.0))),
        Transaction::new_dispute(1,1)]),
    case::resolved_dispute(
        vec![Transaction::new_deposit(1, 1, Some(dec!(1.0))),
        Transaction::new_deposit(2, 2, Some(dec!(2.0))),
        Transaction::new_deposit(1, 3, Some(dec!(2.0))),
        Transaction::new_withdrawal(1, 4, Some(dec!(1.5))),
        Transaction::new_withdrawal(2, 5, Some(dec!(3.0))),
        Transaction::new_dispute(1,1),
        Transaction::new_resolve(1,1)]),
    case::chargeback(
        vec![Transaction::new_deposit(1, 1, Some(dec!(1.0))),
        Transaction::new_deposit(2, 2, Some(dec!(2.0))),
        Transaction::new_deposit(1, 3, Some(dec!(2.0))),
        Transaction::new_withdrawal(1, 4, Some(dec!(1.5))),
        Transaction::new_withdrawal(2, 5, Some(dec!(3.0))),
        Transaction::new_dispute(1,1),
        Transaction::new_chargeback(1,1)])
    )]
    fn test_dispute_flow(transactions: Vec<Transaction>) {
        let result = generate_client_accounts(transactions);

        insta::assert_json_snapshot!(serde_json::json!(result));
    }
}
