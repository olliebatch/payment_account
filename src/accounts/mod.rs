mod generate;
mod write;

pub use generate::*;
pub use write::*;

use crate::transactions::{Transaction, TransactionType};
use rust_decimal::Decimal;
use rust_decimal_macros::dec;

#[derive(serde::Deserialize, serde::Serialize, Debug, PartialEq, Clone)]
pub struct ExternalClientAccount {
    client: u16,
    available: Decimal,
    held: Decimal,
    total: Decimal,
    locked: bool,
}

impl From<&ClientAccount> for ExternalClientAccount {
    fn from(client: &ClientAccount) -> Self {
        ExternalClientAccount {
            client: client.client,
            available: client.available,
            held: client.held,
            total: client.total,
            locked: client.locked,
        }
    }
}

#[derive(serde::Deserialize, serde::Serialize, Debug, PartialEq, Clone)]
pub struct ClientAccount {
    client: u16,
    available: Decimal,
    held: Decimal,
    total: Decimal,
    locked: bool,
    transactions: Vec<Transaction>,
}

impl ClientAccount {
    pub fn new(client_id: u16) -> Self {
        // start a new account with 0 balance and unlocked
        ClientAccount {
            client: client_id,
            available: dec!(0.0),
            held: dec!(0.0),
            total: dec!(0.0),
            locked: false,
            transactions: vec![],
        }
    }

    #[cfg(test)]
    pub fn new_for_testing(client_id: u16, balance: Decimal) -> Self {
        ClientAccount {
            client: client_id,
            available: balance,
            held: dec!(0.0),
            total: balance,
            locked: false,
            transactions: vec![],
        }
    }

    pub fn deposit_to_account(mut self, transaction: Transaction) -> Self {
        if self.is_locked() {
            return self;
        }
        // assume safe to unwrap as deposit
        let amount = transaction.amount.unwrap();
        let new_avail = self.available + amount;
        let total = self.total + amount;

        self.transactions.push(transaction);
        ClientAccount {
            available: new_avail,
            total,
            transactions: self.transactions,
            ..self
        }
    }

    pub fn withdraw_from_account(mut self, transaction: Transaction) -> Self {
        if self.is_locked() {
            return self;
        }
        // assume that an amount is always provided for a withdrawal
        let amount = transaction.amount.unwrap();

        if amount > self.available {
            return self;
        }
        let new_avail = self.available - amount;
        let total = self.total - amount;
        self.transactions.push(transaction);
        ClientAccount {
            available: new_avail,
            total,
            ..self
        }
    }

    fn find_valid_transaction(&self, trx_id: u32) -> Option<&Transaction> {
        self.transactions.iter().find(|&transaction| {
            transaction.tx == trx_id
                && (transaction.payment_type == TransactionType::Withdrawal
                    || transaction.payment_type == TransactionType::Deposit)
        })
    }

    pub fn dispute_to_account(mut self, transaction: Transaction) -> Self {
        let existing_trx = self.find_valid_transaction(transaction.tx);

        if let Some(trx) = existing_trx {
            let trx_amount = trx.amount.unwrap();
            let new_avail = self.available - trx_amount;
            let held_funds = self.held + trx_amount;
            self.transactions.push(transaction);
            return ClientAccount {
                available: new_avail,
                held: held_funds,
                transactions: self.transactions,
                ..self
            };
        } else {
            return self;
        }
    }

    pub fn resolve_to_account(mut self, transaction: Transaction) -> Self {
        let existing_trx = self.find_valid_transaction(transaction.tx);

        if let Some(trx) = existing_trx {
            let trx_amount = trx.amount.unwrap();
            let new_avail = self.available + trx_amount;
            let held_funds = self.held - trx_amount;
            self.transactions.push(transaction);
            return ClientAccount {
                available: new_avail,
                held: held_funds,
                transactions: self.transactions,
                ..self
            };
        } else {
            return self;
        }
    }

    pub fn chargeback_to_account(mut self, transaction: Transaction) -> Self {
        let existing_trx = self.find_valid_transaction(transaction.tx);

        if let Some(trx) = existing_trx {
            let trx_amount = trx.amount.unwrap();
            let held_funds = self.held - trx_amount;
            let total = self.total - trx_amount;
            self.transactions.push(transaction);
            let account = ClientAccount {
                total,
                held: held_funds,
                transactions: self.transactions,
                ..self
            };
            account.lock_account()
        } else {
            return self;
        }
    }

    pub fn is_locked(&self) -> bool {
        self.locked
    }

    pub fn lock_account(self) -> Self {
        ClientAccount {
            locked: true,
            ..self
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::accounts::ClientAccount;
    use crate::transactions::Transaction;
    use rstest::rstest;
    use rust_decimal_macros::dec;

    #[rstest()]
    fn add_deposit_test() {
        let client_account = ClientAccount::new(1);
        let transaction = Transaction::new_deposit(1, 1, Some(dec!(1.5)));

        let updated = client_account.deposit_to_account(transaction);
        assert_eq!(updated.available, dec!(1.5));
        assert_eq!(updated.total, dec!(1.5))
    }

    #[rstest()]
    fn withdraw_funds_test() {
        let client_account = ClientAccount::new_for_testing(1, dec!(5.0));
        let transaction = Transaction::new_withdrawal(1, 1, Some(dec!(1.5)));

        let updated = client_account.withdraw_from_account(transaction);
        assert_eq!(updated.available, dec!(3.5));
        assert_eq!(updated.total, dec!(3.5))
    }

    #[rstest()]
    fn withdraw_funds_more_than_avail() {
        let client_account = ClientAccount::new_for_testing(1, dec!(2.0));
        let transaction = Transaction::new_withdrawal(1, 1, Some(dec!(3.0)));

        let updated = client_account.withdraw_from_account(transaction);
        assert_eq!(updated.available, dec!(2.0));
        assert_eq!(updated.total, dec!(2.0))
    }

    #[rstest()]
    fn dispute_funds_avail() {
        let client_account = ClientAccount::new_for_testing(1, dec!(2.0));
        let transaction = Transaction::new_deposit(1, 1, Some(dec!(3.0)));
        let dispute = Transaction::new_dispute(1, 1);
        let updated = client_account.deposit_to_account(transaction);
        let disputed = updated.dispute_to_account(dispute);

        assert_eq!(disputed.available, dec!(2.0));
        assert_eq!(disputed.held, dec!(3.0));
        assert_eq!(disputed.total, dec!(5.0))
    }
    #[rstest()]
    fn resolve_funds_avail() {
        let client_account = ClientAccount::new_for_testing(1, dec!(2.0));
        let transaction = Transaction::new_deposit(1, 1, Some(dec!(3.0)));
        let dispute = Transaction::new_dispute(1, 1);
        let resolve = Transaction::new_resolve(1, 1);
        let updated = client_account.deposit_to_account(transaction);
        let disputed = updated.dispute_to_account(dispute);
        let resolved = disputed.resolve_to_account(resolve);

        assert_eq!(resolved.available, dec!(5.0));
        assert_eq!(resolved.held, dec!(0.0));
        assert_eq!(resolved.total, dec!(5.0))
    }

    #[rstest()]
    fn chargeback_lock_account() {
        let client_account = ClientAccount::new_for_testing(1, dec!(2.0));
        let transaction = Transaction::new_deposit(1, 1, Some(dec!(3.0)));
        let dispute = Transaction::new_dispute(1, 1);
        let chargeback_trx = Transaction::new_chargeback(1, 1);
        let updated = client_account.deposit_to_account(transaction);
        let disputed = updated.dispute_to_account(dispute);
        let chargeback = disputed.chargeback_to_account(chargeback_trx);

        assert_eq!(chargeback.available, dec!(2.0));
        assert_eq!(chargeback.held, dec!(0.0));
        assert_eq!(chargeback.total, dec!(2.0));
        assert_eq!(chargeback.locked, true);
    }

    #[rstest()]
    fn lock_account() {
        let client_account = ClientAccount::new_for_testing(1, dec!(2.0));

        let account = client_account.lock_account();
        assert_eq!(account.locked, true);
    }
}
