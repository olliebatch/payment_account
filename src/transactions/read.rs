use std::fs::File;

use crate::config::Config;
use crate::transactions::Transaction;
use crate::BoxError;

pub fn read_transaction_file(config: &Config) -> Result<Vec<Transaction>, BoxError> {
    let file = File::open(&config.transaction_file)?;
    let mut rdr = csv::Reader::from_reader(file);

    let mut transactions: Vec<Transaction> = vec![];
    for result in rdr.records() {
        let record = result?;
        let transaction: Transaction = record.deserialize(None)?;
        transactions.push(transaction);
    }

    Ok(transactions)
}
