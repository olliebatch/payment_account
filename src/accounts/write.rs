use crate::accounts::{ClientAccount, ExternalClientAccount};
use crate::BoxError;
use csv::Writer;
use std::collections::HashMap;
use std::io;

pub fn write_to_stdout(accounts: HashMap<u16, ClientAccount>) -> Result<(), BoxError> {
    let mut wtr = Writer::from_writer(io::stdout());

    for account in accounts.values() {
        let external_account: ExternalClientAccount = account.into();
        wtr.serialize(external_account)?;
    }

    wtr.flush()?;
    Ok(())
}
