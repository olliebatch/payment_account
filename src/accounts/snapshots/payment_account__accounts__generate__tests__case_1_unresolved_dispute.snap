---
source: src/accounts/generate.rs
expression: "serde_json::json!(result)"

---
{
  "1": {
    "available": 0.5,
    "client": 1,
    "held": 1.0,
    "locked": false,
    "total": 1.5,
    "transactions": [
      {
        "amount": 1.0,
        "client": 1,
        "payment_type": "deposit",
        "tx": 1
      },
      {
        "amount": 2.0,
        "client": 1,
        "payment_type": "deposit",
        "tx": 3
      },
      {
        "amount": 1.5,
        "client": 1,
        "payment_type": "withdrawal",
        "tx": 4
      },
      {
        "amount": null,
        "client": 1,
        "payment_type": "dispute",
        "tx": 1
      }
    ]
  },
  "2": {
    "available": 2.0,
    "client": 2,
    "held": 0.0,
    "locked": false,
    "total": 2.0,
    "transactions": [
      {
        "amount": 2.0,
        "client": 2,
        "payment_type": "deposit",
        "tx": 2
      }
    ]
  }
}
